### ISSUE TITLE
Can't create correct .epub: failing to read with any reader

### ISSUE THREAD

**Issue by fobdy** [runapp/rwh_epub#1]
1. Cloned repo
2. Zipped it
3. Renamed to .epub

iBooks shows cover but opens with error. Other readers too

**Comment by runapp** [comment]
I can't reproduce your error cause I have no iOS devices. I checked it with EpubReader for Firefox and koreader, both works fine.
Check if you zipped upon the repo directory. The zip file must have META-INF at the top level.

**Comment by fobdy** [comment]
Ok, i'll try EpubReader to reproduce

**Comment by bgschaid** [comment]
I have the same problem (doesn't work on iBook on Mac OS X but on the EpubReader in Firefox). The generated ePub doesn't work on my Kobo-eBook-Reader as well (only shows blank pages).

**Comment by runapp** [comment]
@bgschaid It's wired. Cause you don't provide any error informations and I don't have an iOS device, it would be impossible to fix this.
Besides, I use koreader on my Kobo. I'll try if it renders fine on nickel.

**Comment by bgschaid** [comment]
OK. Checked with `file` on Linux. That identified it as a `ZIP` while other EPubs were identified as `ePub`. The problem here seems to have been that the when zipping it should be to make sure that

1. `mimetype`is the first file in the zip
2. the file is not compresses
3. no extended information is added

When doing that `file` properly identifies it as a EPub and the Atril Document Reader on Linux can read it (couldn't do it before). iBook on Mac still says that the file is corrupt (no explanation) and on the Kobo I still only get the table of contents and 1200+ empty pages.

Now ran `epubchecker` on the file and get
```
Epubcheck Version 1.2

ERROR: RealWorldHaskell.epub/OEBPS/advanced-library-design-building-a-bloom-filter.html(2): element "title" not allowed here; expected the element end-tag
ERROR: RealWorldHaskell.epub/OEBPS/advanced-library-design-building-a-bloom-filter.html(2): element "link" not allowed here; expected the element end-tag
ERROR: RealWorldHaskell.epub/OEBPS/advanced-library-design-building-a-bloom-filter.html(2): attribute "tppabs" not allowed here; expected attribute "charset", "class", "dir", "hreflang", "id", "media", "rev", "style", "title", "type" or "xml:lang"
ERROR: RealWorldHaskell.epub/OEBPS/advanced-library-design-building-a-bloom-filter.html(2): element "meta" not allowed here; expected the element end-tag
```
and approximately 400 similar lines. Seems like the problem is that some ePub-readers are not tolerant enough to accept these problems

**Comment by runapp** [comment]
@bgschaid Thanks for your exploration on this. I'm only grabbing the whole website of RHW and generated the content for it. Maybe it needs some modification to meet the epub standard. Sigil can do it easily, except it will rearrange file structures and make some links invalid. Maybe opening with Sigil and save for another epub file could help. For now I'll add some comment in README.md .
If you have succeeded with some way, please upload your epub file here. I'll include that into the release.

**Comment by bgschaid** [comment]
OK. Sigil makes it better. iBook on the Mac now displays the ePub. The Kobo also does it but the text seems to have shifted to the right and approximately 4 letters on the right are "outside". Never have seen that.

Ah. And another strange thing: epubchecker now reports 1700 problems

**Comment by runapp** [comment]
@bgschaid Seems Sigil and epubchecker aren't in the same mind.
Would you please upload the epub file somewhere?

**Comment by bgschaid** [comment]
Hi @runapp 
I had a look at the warnings that epubchecker gave and wrote me a little python-script that changed the stuff it objected to (unsupported attributes etc, JavaScript, External links) in the HTML-files. Then I had to clean up the CSS a bit (fixed width for the text body is something eBook-Readers don't like). Now I've got a epub that works on my Kobo and on Apples iBook-reader (and epubchecker has no objections): https://keybase.pub/bgschaid/RealWorldHaskell.epub
Interested in the scripts?
